govdelivery/tests/test_api.py

```diff
@@ -1,76 +1,144 @@
+class TestClientUtilities(unittest.TestCase):
+    def test_get_full_url_to_call(self):
+        # Store and unset existing GOVDELIVERY_BASE_URL env var, if it is set.
+        if os.environ.get('GOVDELIVERY_BASE_URL'):
+            current_govdelivery_base = os.environ['GOVDELIVERY_BASE_URL']
+            del os.environ['GOVDELIVERY_BASE_URL']
 
+        url_to_call = get_full_url_to_call('/path')
+        self.assertEqual(url_to_call, 'https://api.govdelivery.com/path')
 
-def require_auth_vars():
-    require_env_vars(['GOVDELIVERY_USER',
-                      'GOVDELIVERY_PASSWORD',
-                      'GOVDELIVERY_ACCOUNT_CODE'])
+        url_to_call = get_full_url_to_call(
+            '/path',
+            'https://passed-base.govdelivery.com'
+        )
+        self.assertEqual(
+            url_to_call,
+            'https://passed-base.govdelivery.com/path'
+        )
 
+        os.environ['GOVDELIVERY_BASE_URL'] = 'https://env-base.govdelivery.com'
+        url_to_call = get_full_url_to_call('/path')
+        self.assertEqual(url_to_call, 'https://env-base.govdelivery.com/path')
 
-def _create_topic():
-    """ Create a test topic and return [TOPIC_CODE] """
-    account_code = os.environ['GOVDELIVERY_ACCOUNT_CODE']
-    payload = """
-        <topic>
-            <name>Test Topic 2</name>
-            <short-name>Test Topic 2</short-name>
-        </topic>
-        """
-    path = ''.join([
-        '/api/account/',
-        account_code,
-        '/topics.xml'])
-    response = api.authenticated_api_call(path, 'post', payload)
-    root = ET.fromstring(response.text)
-    codes = []
-    codes.append(root.find('to-param').text)
-    return codes
+        try:
+            # Was current_govdelivery_base assigned earlier?
+            current_govdelivery_base
+        except NameError:
+            # No: delete the temporary GOVDELIVERY_BASE_URL env var.
+            del os.environ['GOVDELIVERY_BASE_URL']
+        else:
+            # Yes: Reassign GOVDELIVERY_BASE_URL to what it started out as.
+            os.environ['GOVDELIVERY_BASE_URL'] = current_govdelivery_base
 
+    def test_authenticated_session(self):
+        session = authenticated_session('username', 'password')
+        self.assertEqual(session.auth, ('username', 'password'))
 
-def _delete_topic(codes):
-    """ Delete topic with TOPIC_CODE """
-    account_code = os.environ['GOVDELIVERY_ACCOUNT_CODE']
-    path = ''.join([
-        '/api/account/',
-        account_code,
-        '/topics/'])
-    for code in codes:
-        response = api.authenticated_api_call(path + code + '.xml', 'delete')
+    def test_authenticated_api_call(self):
+        # There is no meaningful way to directly test this method, and it
+        # should get covered anyway by testing the other methods that call it.
+        pass
 
 
-class TestAuthentication(unittest.TestCase):
+class TestGovDelivery(unittest.TestCase):
+    def setUp(self):
+        self.gd = GovDelivery()
```

Comment:
This fails if `GOVDELIVERY_USER` isn't specified in the environment.

Reply:
Whoops. Thanks for catching that. Would you recommend requiring it for testing or trying to not require it for creating an instance of the class?

Reply:
I would pass a username and password in as parameters when creating the class. You could also pass in the account code; you define `self.account_code` but then don't use it anywhere.

```py
self.gd = GovDelivery(username='username', password='password', account_code='XYZ')
```

Reply:
`self.account_code` is used in the `translate_path` function here: https://github.com/cfpb/govdelivery/blob/2cdf6ce9d3891741a1db69edff84af1b71c472b0/govdelivery/api.py#L76-L77

But that may not be the most straightforward way to handle that.

Reply:
Ah, right, I misread what [this line](https://github.com/cfpb/govdelivery/blob/70402af56f7ec23bf4b7945da7cd4392b29c0718/govdelivery/tests/test_api.py#L61) was doing. From the test POV, it feels more straightforward to initialize the `GovDelivery` object with the code than to set the internal `.account_code` variable. This would keep better separation of concerns because then you're initializing/testing the public API of that object instead of having to know about its internals.

Reply:
Gotcha, thanks.

Reply:
Fixed in 88ee0a6